app/models/grda_warehouse/vispdat.rb

```diff
@@ -0,0 +1,252 @@
+      number_of_times.validates :inpatient, unless: -> { inpatient_refused? }
+      number_of_times.validates :crisis_service, unless: -> { crisis_service_refused? }
+      number_of_times.validates :talked_to_police, unless: -> { talked_to_police_refused? }
+      number_of_times.validates :jail, unless: -> { jail_refused? }
+    end
+
+    validates :find_location, length: { in: 0..100 }
+    validates :find_time, length: { in: 0..10 }
+    validates :when_answer, inclusion: { in: when_answers.keys }, allow_blank: true
+    validates :phone, format: { with: US_PHONE_NUMBERS }
+    validates_email_format_of :email
+
+    # Require the refused checkbox to be checked if no answer given
+    # Require an answer if the refused checkbox not checked.
+    %w(
+      years_homeless
+      episodes_homeless
+      emergency_healthcare
+      ambulance
+      inpatient
+      crisis_service
+      talked_to_police
+      jail
+    ).each do |field|
+      # if both blank, indicate that refused must be checked
+      validates [field, '_refused'].join.to_sym, presence: { message: 'should be checked if refusing to answering' }, if: -> { send(field.to_sym).blank? }
+
+      # if both blank, indicate a value is needed
+      validates field.to_sym, presence: { message: 'please enter a value or mark refused' }, if: -> { send([field, '_refused?'].join.to_sym).blank? }
+
+      # if refused checked and answer given
+      validates field.to_sym, absence: { message: 'cannot have an entry if refusing to answer' }, if: -> { send([field, '_refused?'].join.to_sym) }
+    end
+
+    ####################
+    # Callbacks
+    ####################
+    before_save :calculate_score
+
+    def calculate_score
+      self.score = pre_survey_score +
+      history_score +
+      risk_score +
+      social_score +
+      wellness_score
+    end
+    def calculate_score!
+      calculate_score
+      save
+    end
+
+    def calculate_recommendation
+      self.recommendation = case score
+      when 0..3
+        "No Housing Intervention"
+      when 4..7
+        "An Assessment for Rapid Re-Housing"
+      when 8..Float::INFINITY
+        "An Assessment for Permanent Supportive Housing/Housing First"
+      else
+        "Invalid Score"
+      end
+    end
+    def calculate_recommendation!
+      calculate_recommendation
+      save
+    end
+
+    def version
+      2
+    end
+
+    def self.language_options
+      language_answers.map do |k,_|
+        [k.split('_').last.capitalize, k]
+      end
+    end
+
+    def self.sleep_options
+      sleep_answers.map do |k,_|
+        [k.split('_').drop(1).join(' ').titleize, k]
+      end
+    end
+
+    def self.when_options
+      when_answers.map do |k,_|
+        [k.capitalize, k]
+      end
+    end
+
+    def self.options_for enum
+      ['Yes', 'No', 'Refused'].zip self.send(enum).keys
+    end
+
+    def full_name
+      [first_name, last_name].join ' '
+    end
+
+    def age
```

Comment:
That's a new mechanism for calculating age to me.  We went through a bunch (though I think this one might be the cleanest).  That said, since we're using the DOB from the client, we should probably just have this call `GrdaWarehouse::Hud::Client#age`